crystalfontz/dbus/client.py

```diff
@@ -0,0 +1,347 @@
+
+
+@config.command()
+@click.argument("name")
+@click.argument("value")
+@async_command
+@pass_config
+@click.pass_obj
+async def set(obj: Obj, staged: StagedConfig, name: str, value: str) -> None:
+    """
+    Set a parameter in the configuration file.
+    """
+
+    try:
+        run_config_command(obj, staged, ["set", name, value])
+    except ValueError as exc:
+        echo(str(exc))
+        sys.exit(1)
+    else:
+        staged.reload_target()
+    finally:
+        if staged.dirty:
+            warn_dirty()
+
+
+@config.command()
+@click.argument("name")
+@async_command
+@pass_config
+async def unset(staged: StagedConfig, name: str) -> None:
+    """
+    Unset a parameter in the configuration file.
+    """
+    try:
+        staged.unset(name)
+    except ValueError as exc:
+        echo(str(exc))
+        sys.exit(1)
+    else:
+        staged.to_file()
+    finally:
+        if staged.dirty:
+            warn_dirty()
+
+
+@config.command()
+@click.option("--baud/--no-baud", default=True, help="Detect baud rate")
+@click.option(
+    "--device/--no-device", default=True, help="Detect device model and versions"
+)
+@click.option(
+    "--save/--no-save", default=True, help="Whether or not to save the configuration"
+)
+@async_command
+@pass_client
+@pass_config
+@click.pass_obj
+async def detect(
+    obj: Obj,
+    staged: StagedConfig,
+    client: DbusClient,
+    baud: bool,
+    device: bool,
+    save: bool,
+) -> None:
+    """
+    Detect device versions and baud rate.
+    """
+
+    baud_rate = -1
+    model = "<unknown>"
+    hardware_rev = "<unknown>"
+    firmware_rev = "<unknown>"
+
+    if baud:
+        baud_rate = await client.detect_baud_rate()
+
+    if device:
+        model, hardware_rev, firmware_rev = await client.detect_device(
+            none_timeout, none_retry_times
+        )
+
+    if save:
+        try:
+            run_config_command(obj, staged, ["set", "baud_rate", str(baud_rate)])
+            run_config_command(obj, staged, ["set", "model", model])
+            run_config_command(obj, staged, ["set", "hardware_rev", hardware_rev])
+            run_config_command(obj, staged, ["set", "firmware_rev", firmware_rev])
+        except ValueError as exc:
+            echo(str(exc))
+            sys.exit(1)
+        else:
+            staged.reload_target()
+        finally:
+            if staged.dirty:
+                warn_dirty()
+
+
+@main.command(help="0 (0x00): Ping command")
+@click.argument("payload", type=BYTES)
+@async_command
+@pass_client
+async def ping(client: DbusClient, payload: bytes) -> None:
+    pong = await client.ping(payload, none_timeout, none_retry_times)
+    echo(pong)
+
+
+@main.command(help="1 (0x01): Get Hardware & Firmware Version")
+@async_command
+@pass_client
+async def versions(client: DbusClient) -> None:
+    model, hardware_rev, firmware_rev = await client.versions(
+        none_timeout, none_retry_times
+    )
+    echo(f"{model}: {hardware_rev}, {firmware_rev}")
```

Comment:
This works fine for now. But this is going to get unwieldy *real* fast with complex return types, with a lot of logic duplicated from responses.

I think a refactor is needed:

- An `__init__` that allows the arguments straight
- A `cls.from_bytes` that does the byte mappings `__init__` does today

This would allow mapping to continue occuring entirely within the dbus code - I think that's right - but would also allow for constructing Response objects for the purposes of printing.